src/index.js

```diff
@@ -183,22 +183,70 @@ export default function (babel) {
 
 
   // HELPER FUNCTIONS
+  function isFunctionDeclaration(node) {
+    return node && (t.is("FunctionDeclaration", node) || node.type === "NamedArrowDeclaration");
+  }
 
-  function transformTerminalExpressionsIntoArrPush(path, arrId) {
-    path.resync(); // not sure if this is necessary... c/p from addImplicitReturns
+  function transformTails(path, allowLoops, getNewNode) {
+    path.resync();
 
-    let completionRecords = path.get("body").getCompletionRecords();
-    for (let targetPath of completionRecords) {
-      if (!targetPath.isExpressionStatement()) continue;
+    const tailPaths = getTailExpressions(path.get("body"), allowLoops);
+    for (const tailPath of tailPaths) {
+      if (!tailPath) continue;
 
-      let arrPush = t.callExpression(
-        t.memberExpression(arrId, t.identifier("push")),
-        [targetPath.node.expression]
-      );
-      targetPath.replaceWith(arrPush);
+      if (tailPath.isExpressionStatement()) {
+        // TODO: add linting to discourage
+        if (tailPath.get("expression").isAssignmentExpression()) {
+          tailPath.insertAfter(getNewNode(tailPath.node.expression.left, tailPath));
+        } else {
+          tailPath.replaceWith(getNewNode(tailPath.node.expression, tailPath));
+        }
+      } else if (tailPath.isVariableDeclaration()) {
+        // TODO: handle declarations.length > 1
+        // TODO: add linting to discourage
+        tailPath.insertAfter(getNewNode(tailPath.node.declarations[0].id, tailPath));
+      } else if (isFunctionDeclaration(tailPath.node)) {
+        // Need to transform exprs to statements since this is block context.
+        const nextNode = getNewNode(tailPath.node.id, tailPath);
+        if (t.is("Expression", nextNode)) {
```

Comment:
Hmm, wouldn't this be necessary for the VariableDeclaration codepath too?

I guess the test case for that would be 
```
[for idx i in Array(10): 
  const x = f(i);
]
```
which reminds me we should definitely have a test for 
```
[for idx i in Array(10): 
  x = f(i);
]
```
which I have a hunch may not do what we want it to. 

Though I'm not sure in either case what we really want here. 

With implicit returns, it can be kind of nice to annotate your return value with a name – or at least, it can be annoying when doing so results in implicit returns turning off or breaking. I guess the same logic applies for comprehensions, so I think we want the same behavior, but it's a little weird (and should almost definitely be a linting error). 

If you just wanna get this shipped for now though I'm happy to merge and add the above in a follow-on